src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -135,56 +206,17 @@ void ZooKeeper::init(ZooKeeperArgs args_)
         {
             tryConnectSameAZKeeper();
         }
-        // TODO: reduce nesting level.
         else
         {
-            Coordination::ZooKeeper::Nodes nodes;
-            nodes.reserve(args.hosts.size());
-
+            bool dns_error = false;
             /// Shuffle the hosts to distribute the load among ZooKeeper nodes.
-            std::vector<ShuffleHost> shuffled_hosts;
-            shuffled_hosts = shuffleHosts();
+            std::vector<ShuffleHost> shuffled_hosts = shuffleHosts(dns_error);
+            if (shuffled_hosts.empty())
+                throwWhenNoHostAvailable(dns_error);
 
-            // TODO: add DNS check as well.
-            bool dns_error = false;
+            Coordination::ZooKeeper::Nodes nodes;
             for (auto & host : shuffled_hosts)
-            {
-                auto & host_string = host.host;
-                try
-                {
-                    const bool secure = startsWith(host_string, "secure://");
-
-                    if (secure)
-                        host_string.erase(0, strlen("secure://"));
-
-                    /// We want to resolve all hosts without DNS cache for keeper connection.
-                    Coordination::DNSResolver::instance().removeHostFromCache(host_string);
-
-                    const Poco::Net::SocketAddress host_socket_addr{host_string};
-                    LOG_TEST(log, "Adding ZooKeeper host {} ({})", host_string, host_socket_addr.toString());
-                    nodes.emplace_back(Coordination::ZooKeeper::Node{host_socket_addr, host.original_index, secure});
-                }
-                catch (const Poco::Net::HostNotFoundException & e)
-                {
-                    /// Most likely it's misconfiguration and wrong hostname was specified
-                    LOG_ERROR(log, "Cannot use ZooKeeper host {}, reason: {}", host_string, e.displayText());
-                }
-                catch (const Poco::Net::DNSException & e)
-                {
-                    /// Most likely DNS is not available now
-                    dns_error = true;
-                    LOG_ERROR(log, "Cannot use ZooKeeper host {} due to DNS error: {}", host_string, e.displayText());
-                }
-            }
-
-            if (nodes.empty())
-            {
-                /// For DNS errors we throw exception with ZCONNECTIONLOSS code, so it will be considered as hardware error, not user error
-                if (dns_error)
-                    throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
-                else
-                    throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
```

Comment:
As you can see, the `Coordination::Error::ZCONNECTIONLOSS` error code here is wrong. I dig a bit, supposed to return bad argument in case of host not found. which is treated as hardware error. but I don't want to introduce diff in this PR.